### DISCUSSION TITLE
Wired trainer interface specification (voltage, frame structure, timing, etc.)

### DISCUSSION THREAD

**Discussion by kobelev-tech** [headtracker/HeadTracker#80] [Q&A] [answered]
Dear sirs, your documentation says: 

> PPM output is by far the most widely supported format for a trainer connection. Almost every radio that has a trainer port will support it.

But there is no any other information, like voltage, format of the frame, timing, etc. Maybe the voltage is the most important as this may burn the transmitter.

I am sure you have this information at least for major brands. Can you please, share what you have?

I am interested in **wired connection** with a cable. My transmitter is old and does not support any radio based modes.

Thanks,
Kirill

**Comment by dlktdr** [comment, marked as answer]
With PPM there isn't a strict standard available. There are systems that use it inverted, different sync pulses, channel count and frame lengths or even a variable frame length. Even the voltage isn't well specified. Most manufactures only state that it supports PPM, very little information on the details on the actual signal required. From the various pieces of information I have found the Frame length is fixed at usually 20ms-22.5ms, 8ch's with a 300-400us sync pulse. Just from random internet links. and the majority of the signal levels being 3.3v.

http://www.dronefromchina.com/new/UAV-PWM-signal-and-PPM-signal.html

With the output from the Nano BLE only being 3.3V, which is the lowest possible voltage any processor currently produced will use it's safe to just hook it up and see if it's going to work. With a radio that expects a 5V ppm input, a 3.3v level is also likely to bee seen as high. When it comes to a really old radio (back in the FM days) a 7 volt PPM might be around. but doubt your going to see it now still won't hurt feeding a 3.3v into this, just probably won't pick it up.

BUT If you have a radio that can do PPM in/out on the same pin it's good just to check that there isn't a voltage on the pin that's higher than 3.9v (nrf52840s absolute maximum voltage) when configured as output/PPM slave (The trainees remote). Just incase it ever gets switched in the settings.

The GUI is pretty configurable to tweak all the PPM settings until you get a working result. What radio do you have?

**Reply by kobelev-tech** [reply]
Hi dlktdr, Thank you very much for your prompt response. I am very sorry that I cannot prepare the full response quickly. I will send several responses instead.
I have 2 transmitters: JR F400 and JR x9303. They are both very old. It looks like I cannot use them with your project at all. Nevertheless. Let me share some info with you.
![image](https://user-images.githubusercontent.com/90671453/161138922-6103abaf-a725-4b52-bdde-1a7910ed11f4.png)
Transmitter has analog voltmeter that shows the voltage on the battery. I described its state in the column #4. The values in the column show if the voltmeter needle is pointing left or right.
The last column shows what is going on on the line when the cord is inserted. I simply attached it to oscilloscope.
"Pulses" - some PWM sequence with higher level at +0.4 v and lover level at -1.3v. As you can see, this is not 3.3v at all. I do not think that 3.3v will burn the transmitter, although I have not tried this.
"Slowly raising to zero". When I press the Trainer switch, voltage on the line drops to -4v, and then slowly raises. In about 10 seconds reaches zero. It is almost certain that the line works as a receiver in this state and voltage on the line is some remaining charge on the high impedance input. Like it or not, it works this way.

I do not see any way to specify the mask of the channels that should be controlled from the slave, while other channels, not in the mask, to be still controlled with the sticks on the transmitter. So, this transmitter will not work with any setup on the Arduino side.

**Reply by dlktdr** [reply]
You might be able add a bit of circuitry to get past that voltage issues but the not being able to selectively set a few channels is a killer. Best of luck.. 
Might be able to intercept the PPM output to the FM module still ;) Although I don't recommend this, too easy for something wrong to kill your radio link and loose a model.

**Reply by kobelev-tech** [reply]
Hi dlktdr. Here is the next part of my study. In this case it is about my second transmitter. This transmitter allows selecting bypassed channels:
![image](https://user-images.githubusercontent.com/90671453/161140224-11ae3f76-9dd7-45e0-a5d9-28277c6444b9.png)
The documentation on the device says the following: 
> The X9303 2.4 contains a Trainer System that allows the instructor to transfer some or all of the primary flight control  functions (Throttle, Aileron, Elevator and Rudder) to the student.

This implies that other channels cannot be transferred, although I am not able to test this explicitly.
Here is the picture of what I see on the oscilloscope with this transmitter:

![image](https://user-images.githubusercontent.com/90671453/161141011-9fed68ec-a3ea-4f1c-af06-50fe33095aea.png)

This is not surprising that it shows groups of 10 pulses for a 9-channel transmitter.
So I am interested on the situation with other manufactures and models of transmitters:
- Do they allow controlling all/other channels? I do not think that it is great idea to dedicate 2 of the 4 main channels to the head tracker.
- Transmitters that I have require constantly pressing the Trainer Switch to accept the input signal. This is not great again. How is this solved with other transmitters? Is this with iron and soldering?

I looked into documentation on the recent Spektrum RC NX10 transmitter. It speaks about wireless trainer and even about FPV head tracking. At the same time it is not speaking about Wi-Fi or Blue Tooth in this section. It is speaking about dedicated trainer receiver and binding instead..

**Reply by dlktdr** [reply]
Yeah that signal looks like pretty standard PPM, fairly low voltage? Your probe is a x1?

I guess the only real way to try is to hook the output from the Nano into it and see. You could easily do a voltage divider on the Nano output to get the same voltage your seeing on the output.

I don't see why overriding a channel 1-4 is going to hurt anything, just re-arrange the receiver. Assuming you can move one of the sticks to a higher channel in the configs.

Open/Edgetx supported radios can customize pretty much everything, you can have ch1 input the trainer output on ch8 to receiver, pretty much whatever arrangement you can think of.

On the Specktrum which also has a similar menu for trainer overrides and switch choice you can force it on  or pick a switch that latches so you don't have to continually hold. You could probably swap that switch with a latching or like you said just solder it.

The Specktrum wireless trainer works over DSM/X same as their radio link. So you can connect any two Specktrums together for trainer. As long as the master has the wireless trainer feature that is. There is a board you can find on eBay that takes the nano PPM output and send it over dsm wirelessly but not directly.
